kedro-telemetry/kedro_telemetry/plugin.py

```diff
@@ -78,10 +80,64 @@ def _get_or_create_uuid() -> str:
         return ""
 
 
+def _get_or_create_project_id(pyproject_path: Path) -> str | None:
+    """
+    Reads a project id from a configuration file or generates and saves a new one if not present.
+    Returns None if configuration file does not exist or does not relate to Kedro.
+    """
+    if pyproject_path.exists():
+        with open(pyproject_path, "r+") as file:
+            pyproject_data = toml.load(file)
+
+            # Check if pyproject related to kedro
+            try:
+                _ = pyproject_data["tool"]["kedro"]
+                try:
+                    project_id = pyproject_data["tool"]["kedro_telemetry"]["project_id"]
+                except KeyError:
+                    project_id = uuid.uuid4().hex
+                    toml_string = (
+                        f'\n[tool.kedro_telemetry]\nproject_id = "{project_id}"\n'
+                    )
+                    file.write(toml_string)
```

Comment:
I 100% agree, there are two reasons why I added it to the separate section for now. 

The first is because `toml.dump` changes formatting like so:
```
[project.optional-dependencies]
docs = [
    "docutils<0.18.0",
    "sphinx~=3.4.3",
    "sphinx_rtd_theme==0.5.1",
    "nbsphinx==0.8.1",
    "sphinx-autodoc-typehints==1.11.1",
    "sphinx_copybutton==0.3.1",
    "ipykernel>=5.3, <7.0",
    "Jinja2<3.1.0",
    "myst-parser~=0.17.2",
]
```

->

```
[project.optional-dependencies]
docs = [ "docutils<0.18.0", "sphinx~=3.4.3", "sphinx_rtd_theme==0.5.1", "nbsphinx==0.8.1", "sphinx-autodoc-typehints==1.11.1", "sphinx_copybutton==0.3.1", "ipykernel>=5.3, <7.0", "Jinja2<3.1.0", "myst-parser~=0.17.2",]
```
Which we do not like but I can replace it with custom file parsing to keep it proper.

2. The main one is because we validate expected keys in the `[tool.kedro]` section and raise an error in case of mismatch: https://github.com/kedro-org/kedro/blob/27f5405cefd6701ffac4c6243030486fb7d3c942/kedro/framework/startup.py#L107

So, to make it properly, we need changes on the framework side: possibly updating `kedro new`, as we discussed above. The question is whether we want to merge it like this for now or wait until we apply updates on the side of the framework.

@merelcht, @astrojuanlu

Reply:
Ugh that "found unexpected keys" indeed would make any extra key in `[tool.kedro]` unfeasible for older Kedro versions. I don't have an opinion right now on whether it makes sense to allow more keys going forward, but it would not be a solution for our case due to lots of old Kedro projects out there in the wild, I think.

About style preservation, I'm wondering if tomlkit is a better choice here, but again this would introduce an extra dependency for very little gain.

Waiting for @merelcht's signoff but I'm okay moving forward with this.

Reply:
@ElenaKhaustova thanks for explaining! That's unfortunate, but in that case I'd also be in favour to just merge it like this. It's more of a "nice to have" than a must anyway.